### ISSUE TITLE
`Layout/ExtraSpacing` behaves differently with `ParserEngine: parser_prism`

### ISSUE THREAD

**Issue by technicalpickles** [rubocop/rubocop#12982]
I'm seeing some false positives for `Layout/ExtraSpacing` when using prism_parser, especially when there are single quotes involved.

## Expected behavior

With `example.rb`:

```ruby
# frozen_string_literal: true

# false positive
puts "nothing matters up here
    on this line the single quote'n... you won’t believe what two single quotes do"

# true negative
puts "nothing matters up here
    on this line the single quote... you won’t believe what two single quotes do"

# funny enough, if you  repeat the first false negative twice, it comes back around to passing
```

Running `rubocop --only Layout/ExtraSpacing example.rb` should pass.

## Actual behavior

rubocop fails with:

```
Inspecting 1 file
C

Offenses:

example.rb:5:83: C: [Correctable] Layout/ExtraSpacing: Unnecessary spacing detected.
    on this line the single quote'n... you won’t believe what two single quotes does"
                                                                                  ^

1 file inspected, 1 offense detected, 1 offense autocorrectable
```

Removing `ParserEngine: parser_prism` does result in it passing.

## Steps to reproduce the problem

With a .rubocop.yml:

```yaml
AllCops:
  TargetRubyVersion: 3.3
  ParserEngine: parser_prism
  NewCops: disable

Layout/ExtraSpacing:
  Enabled: true
```

And an `example.rb`:

```ruby
# frozen_string_literal: true

# false positive
puts "nothing matters up here
    on this line the single quote'n... you won’t believe what two single quotes do"

# true negative
puts "nothing matters up here
    on this line the single quote... you won’t believe what two single quotes do"

# funny enough, if you  repeat the first false negative twice, it comes back around to passing
```

Run `rubocop --only Layout/ExtraSpacing example.rb`.

## RuboCop version

```
❯ rubocop -V
1.64.1 (using Parser 3.3.2.0, rubocop-ast 1.31.3, running on ruby 3.3.2) [arm64-darwin23]
```

**Comment by Earlopain** [comment]
Fixed that specific example with https://github.com/ruby/prism/pull/3338 but there are still problems with multi-byte characters. For example https://github.com/ruby/prism/issues/2539#issuecomment-2052598076:
```rb
# 👎
<<-RUBY.squish
  a
  b
RUBY
```

```
test.rb:2:8: C: [Correctable] Layout/ExtraSpacing: Unnecessary spacing detected.
<<-RUBY.squish
       ^^^^
```

```patch
 [#<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 0...3>, @text="# 👎", @type=:tCOMMENT>,
  #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 4...11>, @text="<<\"", @type=:tSTRING_BEG>,
- #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 19...23>, @text="  a\n", @type=:tSTRING_CONTENT>,
- #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 23...27>, @text="  b\n", @type=:tSTRING_CONTENT>,
+ #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 16...20>, @text="  a\n", @type=:tSTRING_CONTENT>,
+ #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 20...24>, @text="  b\n", @type=:tSTRING_CONTENT>,
  #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 27...31>, @text="RUBY", @type=:tSTRING_END>,
  #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 11...12>, @text=".", @type=:tDOT>,
  #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 12...18>, @text="squish", @type=:tIDENTIFIER>,
- #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 31...32>, @text="", @type=:tNL>]
+ #<RuboCop::AST::Token:0xXXX @pos=#<Parser::Source::Range code.rb 18...19>, @text="", @type=:tNL>]
```

In prism the offsets are per byte, so it must translate them to be per string length that parser expects. Seems like it isn't doing that here.

**Comment by Earlopain** [comment]
Multibyte characters mixed with strings will also be fixed in the next prism release with https://github.com/ruby/prism/pull/3400

**Comment by Earlopain** [comment]
Not stale...

**Comment by dvandersluis** [comment]
Prism 1.4.0 has been released which should fix this issue. Please upgrade the gem and try again.
